tailnet/controllers.go

```diff
@@ -0,0 +1,361 @@
+	DERPCtrl        DERPController
+	ResumeTokenCtrl ResumeTokenController
+	TelemetryCtrl   TelemetryController
+}
+
+type CloserWaiter interface {
+	Close(context.Context) error
+	Wait() <-chan error
+}
+
+// CoordinatorClient is an abstraction of the Coordinator's control protocol interface from the
+// perspective of a protocol client (i.e. the Coder Agent is also a client of this interface).
+type CoordinatorClient interface {
+	Close() error
+	Send(*proto.CoordinateRequest) error
+	Recv() (*proto.CoordinateResponse, error)
+}
+
+// A CoordinationController accepts connections to the control plane, and handles the Coordination
+// protocol on behalf of some Coordinatee (tailnet.Conn in production).  This is the "glue" code
+// between them.
+type CoordinationController interface {
+	New(CoordinatorClient) CloserWaiter
+}
+
+// DERPClient is an abstraction of the stream of DERPMap updates from the control plane.
+type DERPClient interface {
+	Close() error
+	Recv() (*tailcfg.DERPMap, error)
+}
+
+// A DERPController accepts connections to the control plane, and handles the DERPMap updates
+// delivered over them by programming the data plane (tailnet.Conn or some test interface).
+type DERPController interface {
+	New(DERPClient) CloserWaiter
+}
+
+type ResumeTokenClient interface {
+	RefreshResumeToken(ctx context.Context, in *proto.RefreshResumeTokenRequest) (*proto.RefreshResumeTokenResponse, error)
+}
+
+type ResumeTokenController interface {
+	New(ResumeTokenClient) CloserWaiter
+	Token() (string, bool)
+}
+
+type TelemetryClient interface {
+	PostTelemetry(ctx context.Context, in *proto.TelemetryRequest) (*proto.TelemetryResponse, error)
+}
+
+type TelemetryController interface {
+	New(TelemetryClient)
+}
+
+// ControlProtocolClients represents an abstract interface to the tailnet control plane via a set
+// of protocol clients. The Closer should close all the clients (e.g. by closing the underlying
+// connection).
+type ControlProtocolClients struct {
+	Closer      io.Closer
+	Coordinator CoordinatorClient
+	DERP        DERPClient
+	ResumeToken ResumeTokenClient
+	Telemetry   TelemetryClient
+}
+
+type ControlProtocolDialer interface {
+	// Dial connects to the tailnet control plane and returns clients for the different control
+	// sub-protocols (coordination, DERP maps, resume tokens, and telemetry).  If the
+	// ResumeTokenController is not nil, the dialer should query for a resume token and use it to
+	// dial, if available.
+	Dial(ctx context.Context, r ResumeTokenController) (ControlProtocolClients, error)
+}
+
+// basicCoordinationController handles the basic coordination operations common to all types of
+// tailnet consumers:
+//
+//  1. sending local node updates to the Coordinator
+//  2. receiving peer node updates and programming them into the Coordinatee (e.g. tailnet.Conn)
+//  3. (optionally) sending ReadyToHandshake acknowledgements for peer updates.
+type basicCoordinationController struct {
```

Comment:
I am partial to having interface type assertions on these structs (and the ones in the next PRs), even in cases when it's somewhat obvious what interface they implement - but that might be just me.

Reply:
I'm not sure what you mean.  Like `var _ CoordinationController = &basicCoordinationController` type stuff?

One problem with that is is that is defeats the compiler check that private structs are actually used somewhere.